### ISSUE TITLE
500 Internal Server Error when searching for nonexistent identifier using the REST API

### ISSUE THREAD

**Issue by Silberklinge** [bootlin/elixir#91]
Hello,

I am currently leveraging elixir's REST API to quickly grab the history of a struct's definition across different versions of the Linux kernel. But it seems that using the REST API to search for an identifier that doesn't exist causes an 500 status error. For example, the following works just fine:

```
Request:
https://elixir.bootlin.com/api/ident/linux/tc_cls_flower_offload?version=v4.6

Response:
{"definitions": [{"path": "include/net/pkt_cls.h", "line": 417, "type": "struct"}], "references": [{"path": "drivers/net/ethernet/mellanox/mlx5/core/en_tc.c", "line": "103,311,376", "type": null}, {"path": "drivers/net/ethernet/mellanox/mlx5/core/en_tc.h", "line": "42,44", "type": null}, {"path": "include/linux/netdevice.h", "line": "799", "type": null}, {"path": "include/net/pkt_cls.h", "line": "417", "type": null}, {"path": "net/sched/cls_flower.c", "line": "171,194", "type": null}]}
```

But this does not:

```
Request:
https://elixir.bootlin.com/api/ident/linux/tc_cls_flower_offload?version=v4.5

Response:
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at
 [no address given] to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
<hr>
<address>Apache/2.4.29 (Ubuntu) Server at elixir.bootlin.com Port 80</address>
</body></html>
```

The response should be something sane, like `{"definitions": [],  "references": []}`.

**Comment by michaelopdenacker** [comment]
Hi
Many thanks for this report!
Surprisingly, I don't get anything in the server logs.
@carmeli-tamir, how did you get the errors when you developed this script?
Cheers,
Michael.

**Comment by carmeli-tamir** [comment]
I am able to see the error on my machine in /var/log/apache2/error.log, it looks like this: 
`[wsgi:error] [pid 25963:tid 140022401378048] [client 127.0.0.1:60668]   File "falcon/errors.py", line 104, in falcon.errors.HTTPNotFound.__init__`

The falcon API has changed a bit and there is an easy fix, I suggest to return 404 Not Found response, with the following response content:
```
{
    "title": "Identifier not found",
    "description": "Identifier tc_cls_flower_offload not found in project linux version v4.6"
}
```
 IMO returning a 404 is more explicit than return empty results, WDYT?

**Comment by Silberklinge** [comment]
> IMO returning a 404 is more explicit than return empty results, WDYT?

A lack of results isn't an error. Searching for an identifier that doesn't exist is an expected part of user flow.

Besides, an API user that wants to convert an empty list to an error can easily do so on the client side. But an API user that wants to tolerate empty results should not have to write extra code just to convert a 404 message to an empty list.

**Comment by carmeli-tamir** [comment]
It seems that this question has had some long debates, [example1](https://stackoverflow.com/questions/13366730/proper-rest-response-for-empty-table) and [example2](https://stackoverflow.com/questions/11746894/what-is-the-proper-rest-response-code-for-a-valid-request-but-an-empty-data)
Anyway it does seem reasonable to return empty lists. I'll submit a PR soon

**Comment by michaelopdenacker** [comment]
Fixed by PR #92 (thanks to @carmeli-tamir), and deployed on https://elixir.bootlin.com/
Many thanks for your contributions to this improvement.
Michael.
